onnxscript/function_libs/torch_lib/ops/core.py

```diff
@@ -4297,36 +4297,8 @@ def aten_index_put_bool(
     """index_put(Tensor self, Tensor?[] indices, Tensor values, bool accumulate=False) -> Tensor"""
 
     index = op.SequenceAt(indices, 0)  # assume indices only have 1 element
-    # FIXME: ORT ArgMax fails on INT64 input even though ONNX allows it
-    index_int = op.Cast(index, to=INT32.dtype)
-    # if all False, return op.Identity(self)
-    if op.ReduceSum(index_int) == 0:
-        result = self
-    else:
-        # change array([F,F,T,F,F]) to array([2])
-        index = op.ArgMax(index_int)  # assume index only have 1 True
-        # change array([2]) to array([2,2,2,2,2])
-        self_dim_1 = op.Shape(self, start=1, end=2)
-        index_dim_0 = op.Shape(index, start=0, end=1)
-        shape = op.Concat(self_dim_1, index_dim_0, axis=0)
-        new_ind = op.Expand(index, shape)
-        new_ind_t = op.Transpose(new_ind)
-
-        # values must have same rank with input(self)
-        if op.Size(op.Shape(values)) < op.Size(op.Shape(self)):  # type: ignore[operator]
-            values = op.Unsqueeze(values, op.Constant(value_ints=[0]))
-
-        if op.Cast(accumulate, to=BOOL.dtype):
-            zeros = op.Expand(op.Constant(value_float=0.0), op.Shape(self))
-            zeros = op.CastLike(zeros, values)
-            result = op.ScatterElements(zeros, new_ind_t, values)
-            # FIXME: type promotion
-            result = op.CastLike(result, self)
-            result = op.Add(result, self)
-        else:
-            result = op.ScatterElements(self, new_ind_t, values)
-
-    return result
+    # accumulate should be always False, True does not make sense but an assert would be great
+    return op.Where(index, values, self)
```

Comment:
This is much simpler. One problem is that this requires `values` to be broadcastable to `self`. But that may not always be the case. Eg.:
```py
import torch

x = torch.Tensor([0, 1, 2, 3])
print(x)

first_half = x < 2
print(first_half)

y = torch.Tensor([99, 98])
x[first_half] = y

print(x)
```

Reply:
The above produces the output:
```
tensor([0., 1., 2., 3.])
tensor([ True,  True, False, False])
tensor([99., 98.,  2.,  3.])
```

Reply:
It should work in the case we have, where the RHS (values) is a scalar

Reply:
So, it is definitely better than the existing version. So, it is fine to merge it, but I think we may need a more complete solution at some point.

Reply:
What do you mean? When indices are boolean, I don't think we need more. Operator where takes care of the broadcast. When indices are integer, that's another story.

Reply:
The example above uses indices as booleans: I don't think the translation will work for that example.

Reply:
Moved to #1749